File: backend/app/schemas/submission.py
from datetime import datetime

from pydantic import BaseModel, Field, field_validator

from app.models.message import BODY_MAX
from app.models.submission import FILE_REF_MAX, LINK_MAX

MESSAGE_BODY_MIN = 1


class MessageCreate(BaseModel):
    """Payload to create the first message (on submission) or add a message to thread."""

    body: str = Field(..., min_length=MESSAGE_BODY_MIN, max_length=BODY_MAX)


class MessageResponse(BaseModel):
    id: str
    submission_id: str
    sender_id: str
    body: str
    created_at: datetime

    model_config = {"from_attributes": True}


class SubmissionCreate(BaseModel):
    """Payload to submit a solution: mandatory message, optional link."""

    message: str = Field(..., min_length=MESSAGE_BODY_MIN, max_length=BODY_MAX)
    link: str | None = Field(None, max_length=LINK_MAX)
    file_ref: str | None = Field(None, max_length=FILE_REF_MAX)

    @field_validator("link")
    @classmethod
    def link_format(cls, v: str | None) -> str | None:
        if v is None:
            return None
        s = v.strip()
        if not s:
            return None
        if not (s.startswith("http://") or s.startswith("https://")):
            raise ValueError("Link must start with http:// or https://")
        return s


class SubmissionResponse(BaseModel):
    id: str
    project_id: str
    learner_id: str
    link: str | None
    file_ref: str | None
    created_at: datetime
    coherent: bool | None
    message_count: int = 0
    unread_count: int = 0  # For current viewer (learner or owner)

    model_config = {"from_attributes": True}


class SubmissionWithMessagesResponse(SubmissionResponse):
    """Submission with full message thread."""

    messages: list[MessageResponse] = []


class SubmissionCoherentUpdate(BaseModel):
    """Owner sets coherence: True = coherent, False = not coherent."""

    coherent: bool

File: frontend/src/types/submission.ts
export type Submission = {
  id: string
  projectId: string
  learnerId: string
  link: string | null
  fileRef: string | null
  createdAt: string
  coherent: boolean | null
  messageCount: number
  unreadCount: number
}

export type SubmissionWithMessages = Submission & {
  messages: Message[]
}

export type Message = {
  id: string
  submissionId: string
  senderId: string
  body: string
  createdAt: string
}

/** API response shape (snake_case from backend) */
export type SubmissionApiResponse = {
  id: string
  project_id: string
  learner_id: string
  link: string | null
  file_ref: string | null
  created_at: string
  coherent: boolean | null
  message_count: number
  unread_count?: number
}

export type MessageApiResponse = {
  id: string
  submission_id: string
  sender_id: string
  body: string
  created_at: string
}

export type SubmissionWithMessagesApiResponse = SubmissionApiResponse & {
  messages: MessageApiResponse[]
}

export function mapSubmissionFromApi(raw: SubmissionApiResponse): Submission {
  return {
    id: raw.id,
    projectId: raw.project_id,
    learnerId: raw.learner_id,
    link: raw.link,
    fileRef: raw.file_ref,
    createdAt: raw.created_at,
    coherent: raw.coherent,
    messageCount: raw.message_count,
    unreadCount: raw.unread_count ?? 0,
  }
}

export function mapMessageFromApi(raw: MessageApiResponse): Message {
  return {
    id: raw.id,
    submissionId: raw.submission_id,
    senderId: raw.sender_id,
    body: raw.body,
    createdAt: raw.created_at,
  }
}

export function mapSubmissionWithMessagesFromApi(
  raw: SubmissionWithMessagesApiResponse
): SubmissionWithMessages {
  return {
    ...mapSubmissionFromApi(raw),
    messages: raw.messages.map(mapMessageFromApi),
  }
}

export type SubmissionCreateInput = {
  message: string
  link?: string | null
  fileRef?: string | null
}

export type MessageCreateInput = {
  body: string
}
